File: dp_labs_and_solutions/dp_labs_and_solutions/dp_matlab_labs_and_solutions/decorator/lab/Barista.m
% Barista.m

classdef Barista < handle
	methods
		function result = makeCoffee(theBarista, aDescription)
			result = []; % when we add exceptions, we can remove this.

			switch aDescription.getSize
				case 1
					if strcmp(aDescription.get(1), 'coffee')
						result = CoffeeWithoutMilkWithoutSugar;
					else
						% throw an exception.
					end
				case 2
					if strcmp(aDescription.get(1), 'coffee')
						if strcmp(aDescription.get(2), 'milk')
							result = CoffeeWithMilkWithoutSugar;
						elseif strcmp(aDescription.get(2), 'sugar')
							result = CoffeeWithoutMilkWithSugar;
						else
							% throw an exception.
						end
					else
						% throw an exception.
					end
				case 3
					if strcmp(aDescription.get(1), 'coffee')
						if strcmp(aDescription.get(2), 'milk')
							if strcmp(aDescription.get(3), 'sugar')
								result = CoffeeWithMilkWithSugar;
							else
								% throw an exception.
							end
						elseif strcmp(aDescription.get(2), 'sugar')
							if strcmp(aDescription.get(3), 'milk')
								result = CoffeeWithMilkWithSugar;
							else
								% throw an exception.
							end
						else
							% throw an exception.
						end
					else
						% throw an exception.
					end
				otherwise
					% throw an exception.
			end
		end
	end
end

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_cpp_labs_and_solutions/decorator/lab/barista.cpp
// barista.cpp

#include <string>
#include <vector>
#include "barista.h"
#include "coffee.h"
#include "coffee_with_milk_with_sugar.h"
#include "coffee_with_milk_without_sugar.h"
#include "coffee_without_milk_with_sugar.h"
#include "coffee_without_milk_without_sugar.h"

coffee * barista::make_coffee(const std::vector<std::string> & a_description)
{
	switch (a_description.size())
	{
		case 1:
			if (a_description[0] == "coffee")
			{
				return new coffee_without_milk_without_sugar();
			}
			else
			{
				// throw an exception.
			}
			break;
		case 2:
			if (a_description[0] == "coffee")
			{
				if (a_description[1] == "milk")
				{
					return new coffee_with_milk_without_sugar();
				}
				else if (a_description[1] == "sugar")
				{
					return new coffee_without_milk_with_sugar();
				}
				else
				{
					// throw an exception.
				}
			}
			else
			{
				// throw an exception.
			}
			break;
		case 3:
			if (a_description[0] == "coffee")
			{
				if (a_description[1] == "milk")
				{
					if (a_description[2] == "sugar")
					{
						return new coffee_with_milk_with_sugar();
					}
					else
					{
						// throw an exception.
					}
				}
				else if (a_description[1] == "sugar")
				{
					if (a_description[2] == "milk")
					{
						return new coffee_with_milk_with_sugar();
					}
					else
					{
						// throw an exception.
					}
				}
				else
				{
					// throw an exception.
				}
			}
			else
			{
				// throw an exception.
			}
			break;
		default:
			// throw an exception.
			break;
	}
	return nullptr; // when we add exceptions, we can remove this.
}
